File: ios/Sources/EmojiPicker/Presenter/DispatchingEmojiPickerPresenter.swift
import Foundation

/// Picks between the native and web-fallback presenters by `options.presentation`, so
/// `EmojiPickerService`'s single concurrency guard covers both paths uniformly - from its point
/// of view this is just another `EmojiPickerPresenter`.
final class DispatchingEmojiPickerPresenter: EmojiPickerPresenter {
    private let nativePresenter: EmojiPickerPresenter
    private let webFallbackPresenter: EmojiPickerPresenter

    init(nativePresenter: EmojiPickerPresenter, webFallbackPresenter: EmojiPickerPresenter) {
        self.nativePresenter = nativePresenter
        self.webFallbackPresenter = webFallbackPresenter
    }

    func present(options: EmojiPickerPresentOptions, completion: @escaping (Result<EmojiPickerResult, EmojiPickerError>) -> Void) {
        let target = options.presentation == "web" ? webFallbackPresenter : nativePresenter
        target.present(options: options, completion: completion)
    }
}

File: android/src/main/java/app/independo/capacitoremojipicker/presenter/DispatchingEmojiPickerPresenter.java
package app.independo.capacitoremojipicker.presenter;

import app.independo.capacitoremojipicker.core.EmojiCloseButtonOptions;
import app.independo.capacitoremojipicker.core.EmojiPickerCallback;

/**
 * Picks between the native and web-fallback presenters by {@code presentation}, so {@code
 * EmojiPickerService}'s single concurrency guard covers both paths uniformly - from its point of
 * view this is just another {@link EmojiPickerPresenter}.
 */
public class DispatchingEmojiPickerPresenter implements EmojiPickerPresenter {

    private final EmojiPickerPresenter nativePresenter;
    private final EmojiPickerPresenter webFallbackPresenter;

    public DispatchingEmojiPickerPresenter(EmojiPickerPresenter nativePresenter, EmojiPickerPresenter webFallbackPresenter) {
        this.nativePresenter = nativePresenter;
        this.webFallbackPresenter = webFallbackPresenter;
    }

    @Override
    public void present(String presentation, boolean dismissOnBackdropTap, EmojiCloseButtonOptions closeButton, EmojiPickerCallback callback) {
        EmojiPickerPresenter target = "web".equals(presentation) ? webFallbackPresenter : nativePresenter;
        target.present(presentation, dismissOnBackdropTap, closeButton, callback);
    }
}
